File: arrays/alphabetBoardPath.py
# https://leetcode.com/problems/alphabet-board-path/

class Solution:
    
    # Runtime: 71 ms, faster than 12.71% of Python3 online submissions for Alphabet Board Path.
    # Memory Usage: 13.8 MB, less than 99.15% of Python3 online submissions for Alphabet Board Path.
    
    def alphabetBoardPath(self, target: str) -> str:
        
        board = {ch:[ind//5, ind%5] for ind , ch in enumerate("abcdefghijklmnopqrstuvwxyz")}
        
        x0, y0 = 0, 0
        
        res = []
        # Order matters L U should come first before R D
        
        for ch in target:
            x, y = board[ch]
            if y < y0: res.append('L' * (y0 - y))
            if x < x0: res.append('U' * (x0 - x))
            if x > x0: res.append('D' * (x - x0))
            if y > y0: res.append('R' * (y - y0))
            res.append("!")
            x0, y0 = x,y
            
        return "".join(res)
            

File: arrays/alphabetBoardPath.js

// Runtime: 84 ms, faster than 61.76% of JavaScript online submissions for Alphabet Board Path.
// Memory Usage: 43.6 MB, less than 26.47% of JavaScript online submissions for Alphabet Board Path.
// /**
//  * @param {string} target
//  * @return {string}
//  */
function alphabetBoardPath(target) {
  const board = ["abcde", "fghij", "klmno", "pqrst", "uvwxy", "z"];
  const boardMap = new Map();
  board.forEach((row, rowIndex) => {
    row.split("").forEach((letter, colIndex) => {
      boardMap.set(letter, [rowIndex, colIndex]);
    });
  });
  const targetAttr = target.split("").map((letter) => boardMap.get(letter));
  const result = [];
  let prevAttr = [0, 0];
  targetAttr.forEach((attr) => {
    const [row, col] = attr;
      const [prevRow, prevCol] = prevAttr;
      if (row === 5 && col === 0) {
          if (row > prevRow) {
              result.push("D".repeat(row - prevRow - 1));
          }
          if (col < prevCol) {
              result.push("L".repeat(prevCol - col));
          }
          if (row > prevRow) {
              result.push("D".repeat(1));
          }
      }
      else {
          if (prevRow > row) {
              result.push("U".repeat(prevRow - row));
          }
          if (row > prevRow) {
              result.push("D".repeat(row - prevRow));
          }
          if (col < prevCol) {
              result.push("L".repeat(prevCol - col));
          }
          if (prevCol < col) {
              result.push("R".repeat(col - prevCol));
          }
      }
    result.push("!");
    prevAttr = attr;
  });
  return result.join("");
}
console.log(alphabetBoardPath("zdz"));
module.exports = alphabetBoardPath;
